### PR TITLE
Merge down updates from upstream

### PR DESCRIPTION
**Pull request by jeffmo**
This was a relatively non-trivial merge, so please scrutinize carefully.

Let's try to start doing these more often (i.e. maybe right after each downstream-fork change?) so that such significant merges can be a bit more iterative.

### CODE REVIEW

**Comment by jeffmo** [comment]
cc @gabelevi

**Comment by sophiebits** [comment]
FWIW git-imerge http://softwareswirl.blogspot.com/2013/05/git-imerge-practical-introduction.html might be useful for these merges in the future.

**Comment by jeffmo** [comment]
(Note: I ran this through our internal esprima_tester and verified the results give us no parse errors or test failures on our internal code)

**Comment by jeffmo** [comment]
@spicyj: Yea, that looks like it might make the process slightly less painful. Ultimately I don't know that it'd have saved a ton of work here though -- there weren't that many upstream changes, but there were quite a few downstream changes made that moved things around in odd slightly non-trivial ways.

**Comment by gabelevi** [comment]
Looks good to me. I checked this against the flow tests to see what broke, and the issues all see to be on Flow's side. They are

1) Previously `([a.a]) => 42` parsed. It was a typo in the test, but should be an error. Esprima fixed this bug, Flow should fix it too.
2) Esprima started tracking computed for MethodDefinitions (well, except for generators :/). Awesome! Flow just needs to be told that Esprima fixed this.
3) Esprima now only parses exports/imports if `sourceType: "module"` is set in the options. Need to update the test runner to turn this on for export/import tests (not globally, since this implicitly turns on strict mode too).

File: esprima.js [hunk 1]
```diff
@@ -5831,13 +5896,53 @@ parseYieldExpression: true, parseAwaitExpression: true
...
             isAsync, typeParameters, tokenValue, returnType,
-            annotationMarker;
+            annotationMarker, propMap;
 
         propType = isStatic ? ClassPropertyType.static : ClassPropertyType.prototype;
 
+        propMap = existingPropNames[propType];
+
         if (generator) {
             return delegate.createMethodDefinition(
```
Comment (gabelevi) [inline esprima.js:5947]: Actually this looks like a merge problem. [Upstream](https://github.com/ariya/esprima/blob/harmony/esprima.js#L4722) it does look like this is passed a 5th argument `computed`

**Comment by jeffmo** [comment]
> Esprima now only parses exports/imports if sourceType: "module" is set in the options. Need to update the test runner to turn this on for export/import tests 

I assume you mean the flow/esprima comparison test runner here, right?

**Comment by eventualbuddha** [comment]
Is there a release planned that incorporates these changes? Thanks!

**Comment by jeffmo** [comment]
I think there's one pending fix for arrows right now, so let's try to push that through in the next few days then I can do a release.

**Comment by jeffmo** [comment]
@eventualbuddha: FYI, just now published `esprima-fb@8001.2001.0-dev-harmony-fb` to npm

**Pull request merged**